### DISCUSSION TITLE
Issue with Mapping a Function in CasADi with Free Variables

### DISCUSSION THREAD

**Discussion by cdhainaut** [casadi/casadi#3752] [Q&A]
Hi everyone,

I'm encountering an issue in CasADi when mapping a function with free variables. Below are three snippets: the first works, the second throws an error when attempting to map the function, and the third shows a workaround but requires explicitly mentioning all variables, which I want to avoid.

**Working Example**

```python
import casadi as cas

opti = cas.Opti()

a = opti.variable()
b = opti.variable()
c = opti.variable()
d = opti.variable()

y = (a - 2 * c) ** 2 * (b - 4 * d) ** 2

opti.subject_to([c >= 3, d >= 4])

f = cas.Function("f", [a, b], [y], {"allow_free": True})

opti.subject_to([a == 1, b == 1])
obj = f(a, b)
opti.minimize(obj)
opti.solver("ipopt")
sol = opti.solve()
```

**Example with Mapping (Error)**

```python
import casadi as cas
import numpy as np

opti = cas.Opti()

a = opti.variable()
b = opti.variable()
c = opti.variable()
d = opti.variable()

y = (a - 2 * c) ** 2 * (b - 4 * d) ** 2

opti.subject_to([c >= 3, d >= 4])

f = cas.Function("f", [a, b], [y], {"allow_free": True})

n_pts = 10
a_s = np.linspace(0, 10, n_pts)
b_s = np.linspace(0, 20, n_pts)
res = f.map(n_pts)(a_s, b_s)

obj = cas.sum1(res)
opti.minimize(obj)
opti.solver("ipopt")
sol = opti.solve()
```

**Error message**
```Function f (0x210d160)
Input 0 (i0): 0
Input 1 (i1): 0
Traceback (most recent call last):
  File "test_cas.py", line 21, in <module>
    res = f.map(n_pts)(a_s, b_s)
  File "casadi.py", line 23373, in __call__
    ret = self.call(args)
  File "casadi.py", line 20031, in call
    return _casadi.Function_call(self, *args)
RuntimeError: Error in Function::call for 'map10_f' [Map] at .../casadi/core/function.cpp:361:
Error in Function::operator() for 'f' [MXFunction] at .../casadi/core/function.cpp:1513:
.../casadi/core/mx_function.cpp:432: Cannot evaluate "f:(i0,i1)->(o0) MXFunction" since variables [opti0_x_3, opti0_x_4] are free.
```

**Example with map working, without free variables**

```python
import casadi as cas
import numpy as np


opti = cas.Opti()

a = opti.variable()
b = opti.variable()
c = opti.variable()
d = opti.variable()

y = (a - 2 * c) ** 2 * (b - 4 * d) ** 2

opti.subject_to([c >= 3, d >= 4])

f = cas.Function("f", [a, b, c, d], [y])

n_pts = 10
a_s = np.linspace(0, 10, n_pts)
b_s = np.linspace(0, 20, n_pts)
res = f.map(n_pts)(a_s, b_s, c, d)

obj = cas.sum1(res)
opti.solver("ipopt")
sol = opti.solve()
```

In the first example, the function f is defined with a and b, while c and d are treated as free variables, which works as expected. In the second example, I attempt to map f over multiple values of a and b, but I encounter an error due to c and d being free variables. The third example resolves this issue by explicitly listing all variables, but this approach can lead to cumbersome and lengthy lists of variables, complicating the code unnecessarily. How can I correctly map the function without explicitly listing all the dependent variables?

The reason I need to achieve this is that in my actual code, the "y" function does not support a and b as vectors or matrices, only as scalars. This approach is a workaround to bypass shape limitations (CasADi supports only matrices and not tensors of arbitrary shapes) without having to refactor the entire codebase.

Thank you for your help!

**Comment by cdhainaut** [comment]
Anyone?

**Reply by snDevX** [reply]
i would try and list all my free variables as a single parameter vector. 

That way you will pass only one additional entry in the list of variables which includes all your free variables.

But there could be a better approach
